File: app/src/main/java/com/github/moko256/twitlatte/database/CachedIdListSQLiteOpenHelper.java
/*
 * Copyright 2015-2019 The twitlatte authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.moko256.twitlatte.database;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.database.sqlite.SQLiteStatement;

import com.github.moko256.latte.client.base.entity.AccessToken;

import java.io.File;
import java.util.List;

import kotlin.collections.ArraysKt;

/**
 * Created by moko256 on 2017/06/08.
 *
 * @author moko256
 */

public class CachedIdListSQLiteOpenHelper extends SQLiteOpenHelper {

    private static final String ID_LIST_TABLE_NAME = "IdList";
    private static final String[] COLUMNS = new String[]{"id"};

    private static final String SEEING_ID_TABLE_NAME = "SeeingId";

    private static final String insertIdListStatement = "insert into " + ID_LIST_TABLE_NAME + " values(?)";
    private static final String deleteIdListStatement = "delete from " + ID_LIST_TABLE_NAME + " where id=?";

    public CachedIdListSQLiteOpenHelper(Context context, AccessToken accessToken, String name) {
        super(context, accessToken != null ? new File(context.getCacheDir(), accessToken.getKeyString() + "/" + name + ".db").getAbsolutePath() : null, null, 2);
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        db.execSQL("create table " + ID_LIST_TABLE_NAME + "(id)");
        db.execSQL("create table " + SEEING_ID_TABLE_NAME + "(id)");
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        if (oldVersion < 2) {
            db.execSQL("drop table IdList");
            db.execSQL("drop table ListViewPosition");
            onCreate(db);
        }
        // else (oldVersion < 3) ...
    }

    public List<Long> getIds() {
        final long[] ids;

        final Cursor c = getReadableDatabase().query(ID_LIST_TABLE_NAME, COLUMNS, null, null, null, null, null);
        final int count = c.getCount();
        ids = new long[count];

        int i = 1;
        while (c.moveToNext()) {
            ids[count - i] = c.getLong(0);
            i++;
        }

        c.close();

        return ArraysKt.asList(ids);
    }

    public void addIds(List<Long> ids) {
        SQLiteDatabase database = getWritableDatabase();
        database.beginTransaction();
        SQLiteStatement insert = database.compileStatement(insertIdListStatement);
        try {
            addIdsInner(insert, ids);
            database.setTransactionSuccessful();
        } finally {
            database.endTransaction();
        }
        insert.close();
    }

    private void addIdsInner(SQLiteStatement statement, List<Long> ids) {
        for (int i = ids.size() - 1; i >= 0; i--) {
            statement.bindLong(1, ids.get(i));
            statement.execute();
        }
    }

    public void insertIds(int bottomPosition, List<Long> ids) {
        List<Long> n = getIds();
        List<Long> d = n.subList(0, bottomPosition);

        SQLiteDatabase database = getWritableDatabase();
        database.beginTransaction();
        SQLiteStatement insert = database.compileStatement(insertIdListStatement);
        SQLiteStatement delete = database.compileStatement(deleteIdListStatement);
        try {
            deleteIdsInner(delete, d);
            addIdsInner(insert, ids);
            addIdsInner(insert, d);
            database.setTransactionSuccessful();
        } finally {
            database.endTransaction();
        }
        delete.close();
        insert.close();
    }

    public void deleteIds(List<Long> ids) {
        SQLiteDatabase database = getWritableDatabase();
        database.beginTransaction();
        SQLiteStatement delete = database.compileStatement(deleteIdListStatement);
        try {
            deleteIdsInner(delete, ids);
            database.setTransactionSuccessful();
        } finally {
            database.endTransaction();
        }
        delete.close();
    }

    private void deleteIdsInner(SQLiteStatement statement, List<Long> ids) {
        for (Long id : ids) {
            statement.bindLong(1, id);
            statement.execute();
        }
    }

    public long getSeeingId() {
        long r;

        Cursor c = getReadableDatabase()
                .query(SEEING_ID_TABLE_NAME, COLUMNS, null, null, null, null, null);
        if (c.moveToNext()) {
            r = c.getLong(0);
        } else {
            r = 0L;
        }
        c.close();

        return r;
    }

    public void setSeeingId(Long i) {
        ContentValues contentValues = new ContentValues(1);
        contentValues.put("id", i);

        SQLiteDatabase database = getWritableDatabase();
        database.beginTransaction();
        try {
            database.delete(SEEING_ID_TABLE_NAME, null, null);
            database.insert(SEEING_ID_TABLE_NAME, null, contentValues);
            database.setTransactionSuccessful();
        } finally {
            database.endTransaction();
        }
    }

}

File: app/src/main/java/com/github/moko256/twitlatte/database/CachedIdListSQLiteOpenHelper.kt
/*
 * Copyright 2015-2019 The twitlatte authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.moko256.twitlatte.database

import android.content.Context
import android.database.DatabaseUtils
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.database.sqlite.SQLiteStatement
import com.github.moko256.latte.client.base.entity.AccessToken
import com.github.moko256.twitlatte.database.utils.read
import com.github.moko256.twitlatte.database.utils.transaction
import com.github.moko256.twitlatte.database.utils.write
import java.io.File

/**
 * Created by moko256 on 2019/12/07.
 *
 * @author moko256
 */
class CachedIdListSQLiteOpenHelper(
    context: Context,
    accessToken: AccessToken?,
    name: String

) : SQLiteOpenHelper(
    context,
    accessToken?.let {
        File(
            context.cacheDir,
            "${it.getKeyString()}/$name.db"
        ).absolutePath
    },
    null,
    2
) {
    companion object {
        private val COLUMNS = arrayOf("id")

        private const val ID_LIST_TABLE_NAME = "IdList"
        private const val SEEING_ID_TABLE_NAME = "SeeingId"

        private const val insertIdListStatement = "insert into $ID_LIST_TABLE_NAME values(?)"
    }

    override fun onCreate(db: SQLiteDatabase) {
        db.execSQL("create table $ID_LIST_TABLE_NAME(id)")
        db.execSQL("create table $SEEING_ID_TABLE_NAME(id)")
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        if (oldVersion < 2) {
            db.execSQL("drop table IdList")
            db.execSQL("drop table ListViewPosition")
            onCreate(db)
        }
        // else (oldVersion < 3) ...
    }

    fun getIds(): List<Long> {
        return read {
            getIdsInner(this)
        }
    }

    private fun getIdsInner(db: SQLiteDatabase): List<Long> {
        val c = db.query(ID_LIST_TABLE_NAME, COLUMNS, null, null, null, null, null)
        val count = c.count
        val ids = LongArray(count)

        var i = count
        while (c.moveToNext()) {
            i--
            ids[i] = c.getLong(0)
        }

        c.close()
        return ids.asList()
    }

    fun insertIdsAtFirst(ids: List<Long>) {
        transaction {
            val insert = compileStatement(insertIdListStatement)

            addIdsInner(insert, ids)

            insert.close()
        }
    }

    fun insertIdsAtLast(ids: List<Long>) {
        transaction {
            val n = getIdsInner(this)

            val insert = compileStatement(insertIdListStatement)

            deleteAllIds(this)
            addIdsInner(insert, ids)
            addIdsInner(insert, n)

            insert.close()
        }
    }

    fun insertIdsAt(index: Int, ids: List<Long>) {
        transaction {
            val n = getIdsInner(this)
            val d = n.subList(0, index)

            val insert = compileStatement(insertIdListStatement)

            deleteTopIds(this, n.size, index)
            addIdsInner(insert, ids)
            addIdsInner(insert, d)

            insert.close()
        }
    }

    fun removeAt(index: Int) {
        write {
            val offset = DatabaseUtils.queryNumEntries(this, ID_LIST_TABLE_NAME).toInt() - index - 1
            execSQL(
                "delete from $ID_LIST_TABLE_NAME where rowid in (select rowid from $ID_LIST_TABLE_NAME limit 1 offset $offset)"
            )
        }
    }

    fun removeFromLast(count: Int) {
        write {
            execSQL(
                "delete from $ID_LIST_TABLE_NAME where rowid in (select rowid from $ID_LIST_TABLE_NAME limit $count)"
            )
        }
    }

    private fun addIdsInner(statement: SQLiteStatement, ids: List<Long>) {
        for (i in ids.indices.reversed()) {
            statement.bindLong(1, ids[i])
            statement.execute()
        }
    }

    private fun deleteAllIds(database: SQLiteDatabase) {
        database.execSQL("delete from $ID_LIST_TABLE_NAME")
    }

    private fun deleteTopIds(database: SQLiteDatabase, listSize: Int, toIndex: Int) {
        val offset = listSize - toIndex
        database.execSQL(
            "delete from $ID_LIST_TABLE_NAME where rowid in (select rowid from $ID_LIST_TABLE_NAME limit $toIndex offset $offset)"
        )
    }


    fun getSeeingId(): Long {
        return read {
            val c = query(SEEING_ID_TABLE_NAME, COLUMNS, null, null, null, null, null)
            val r = if (c.moveToNext()) {
                c.getLong(0)
            } else {
                0L
            }
            c.close()
            r
        }
    }

    fun setSeeingId(id: Long?) {
        transaction {
            delete(SEEING_ID_TABLE_NAME, null, null)
            execSQL("insert into $SEEING_ID_TABLE_NAME values($id)")
        }
    }

}